### DISCUSSION TITLE
Is possible to specify the exe to be used according to the script?

### DISCUSSION THREAD

**Discussion by Momoshikin** [zero-plusplus/vscode-autohotkey-debug#156] [Q&A] [answered]
I was reading the docs and I build my settings like:
```
{
	"folders": [],
	"launch": {
		"version": "0.2.0",
		"configurations": [
			{
				"name": "AutoHotkey Debugger",
				"type": "autohotkey",
				"request": "launch",
				"program": "${file}",
				"usePerfTips": true,
				"useAutoJumpToError": true,
				"useDebugDirective": true
			}
		],
	},

}
```
Is possible to somehow specify what `exe` to use according to the `script`?

I mean:

script `x.ahk` debugger run with `autohotkey.exe`
script `y.ahk` debugger run with `autohotkey1.exe`
script `z.ahk` debugger run with `autohotkey2.exe`

**`Why:`**
I'm working on an obfuscator for the `c++` AutoHotkey source, and as I recompile it often and save different versions when I need to debug a script with one of these `exe`, I need to replace my original AutoHotkey.exe.

Would help a lot to be able to select what `exe´ to use according to the script.

**Comment by zero-plusplus** [comment, marked as answer]
A common solution to this issue is to create as many configurations as there are files with the `program` changed.
```
{
  "name": "AutoHotkey Debugger: x",
  "type": "autohotkey",
  "request": "launch",
  "program": "path/to/x.ahk",
  "usePerfTips": true,
  "useAutoJumpToError": true,
  "useDebugDirective": true
},
{
  "name": "AutoHotkey Debugger: y",
  "type": "autohotkey",
  "request": "launch",
  "program": "path/to/y.ahk",
  "extends": "AutoHotkey Debugger: x",
},
{
  "name": "AutoHotkey Debugger_z",
  "type": "autohotkey",
  "request": "launch",
  "program": "path/to/z.ahk",
  "extends": "AutoHotkey Debugger: x",
}
```
By using the `extends` attribute, common attributes can be used in different settings.

---

Another solution is to set rules for the file structure.

For example, suppose you have a file structure like the following.
```
- root
  - x
    - x.ahk
    - AutoHotkey.exe
  - y
    - y.ahk
    - AutoHotkey.exe
  - z
    - z.ahk
    - AutoHotkey.exe
```

In this case, you can use common settings by using the `${fileWorkspaceFolder}` variable as follows.
```
{
  "name": "AutoHotkey Debugger",
  "type": "autohotkey",
  "request": "${fileWorkspaceFolder}/AutoHotkey.exe",
  "program": "${file}",
  "usePerfTips": true,
  "useAutoJumpToError": true,
  "useDebugDirective": true
},
```
You may be able to solve the issue more flexibly by using other variables. Please check [here](https://code.visualstudio.com/docs/editor/variables-reference) for valid variables.

---

Finally, an example of using `${fileBasenameNoExtension}` is given.
```
- root
  - x.ahk
  - x.exe
  - y.ahk
  - y.exe
  - z.ahk
  - z.exe
```
```
{
  "name": "AutoHotkey Debugger",
  "type": "autohotkey",
  "request": "${fileWorkspaceFolder}/${fileBasenameNoExtension}.exe",
  "program": "${file}",
  "usePerfTips": true,
  "useAutoJumpToError": true,
  "useDebugDirective": true
},
```

**Reply by Momoshikin** [reply]
Got it working with:
```
	"folders": [],
	"launch": {
		"version": "0.2.0",
		"configurations": [
			{
				"name": "AutoHotkey Debugger",
				"type": "autohotkey",
				"request": "launch",
				"program": "${file}",
				"usePerfTips": true,
				"useAutoJumpToError": true,
				"useDebugDirective": true
			},
			{
				"name": "AutoHotkey Debugger: x",
				"type": "autohotkey",
				"request": "launch",
				"runtime": "AutoHotkey1.exe",
				"program": "script.ahk",
				"usePerfTips": true,
				"useAutoJumpToError": true,
				"useDebugDirective": true
			 },

		],
	},
```

But its boring needing to change the current debugger in this option:
![image](https://user-images.githubusercontent.com/92819425/138532807-1f6e54e0-c8ad-4e66-b315-3285319abe43.png)

Isn't possible to specify a way that doesn't involve manually changing this option?

**Reply by zero-plusplus** [reply]
If there was a feature to assign an exe to a file, this issue would be solved.

However, the solution envisioned by VSCode is probably to use variables.

For example, if I tried to solve this issue, I would have to create another attribute to assign an exe to a file.

However, that feature is similar to `runtime` and `program`, which will confuse many users.

So please review the file structure and look for a solution using variables such as the second and third ones above.

If you let us know the file structure, etc., I can work with you to come up with a solution.

---

As an alternative solution, you can use `Ctrl + Alt + F5` to switch debug configurations without using the mouse.
